meja/typechecker.ml

```diff
@@ -0,0 +1,110 @@
+open Core_kernel
```

Comment:
Did you write this using typing rules written down somewhere? If so, it would help with reviewing if you could share that

Reply:
I started with the bidirectional typechecker from the [paper we discussed](https://www.cl.cam.ac.uk/~nk480/bidir.pdf), and then wrote `check_type` below to do the actual checking and unifying to full the gap when I realised it wasn't a good fit.